### ISSUE TITLE
fsremap the error said that that device is not read-only, but the device is read-only

### ISSUE THREAD

**Issue by WeiweiDuan** [cosmos72/fstransform#33]
Hi,

I tried to use fsremap to convert the loop file into the device, but I got the error showing in the following screenshot. Any idea about how to solve it. Thanks!
![fsremap_error](https://user-images.githubusercontent.com/9730105/91668397-4bd41780-eac1-11ea-9af9-1f43ec6707db.jpeg)

**Comment by cosmos72** [comment]
The commands are correct, but it seems that `mount -r` does not work as expected on your `/dev/loop0`.
Which filesystem does it contain?

**Comment by WeiweiDuan** [comment]
Thanks for your response. The filesystem is XFS. I tried to use fstransform command to convert XFS into EXT4, but the /dev/loop0 stored as EXT4 is empty, and the original data converted as loop-file with XFS filesystem.

**Comment by cosmos72** [comment]
Update: there is a mismatch in the paths you're using.

You mount `/dev/loop0` to the directory `/tmp/weiwei2`, but then you attempt to use a file `/tmp/fstransform.mount.23404/.fstranform.loop.26458` which is **not** inside such mount directory.

At least one of the paths you are using is wrong.
Maybe an `ls -al /tmp/weiwei2` (while `/dev/loop0` is mounted there) would help understanding what are the correct paths to use.

**Comment by WeiweiDuan** [comment]
Let me clarify what I want to do. I want to remap the loop-file into original files. According to the instruction in the webpage, I thought I need an empty disk to store the original files. "/dev/loop0" is an empty disk mounted in "/tmp/weiwei2".  The data is in "centos_arya-home" disk mounted in "/tmp/fstransform.mount.23404/.fstranform.loop.26458" directory. Please correct me if I misunderstood the fsremap command. If I'm wrong, do you know how to convert the loop-file back to the original files. Thanks a lot for your help.

**Comment by cosmos72** [comment]
Now it's clearer, thanks.
If your target disk `/dev/loop0` is empty and mounted in `/tmp/weiwei2`, and the data is inside a file `/tmp/fstransform.mount.23404/.fstranform.loop.26458`, there is no need to use `fsremap`: you can just mount the `.fstranform.loop.XYZ` loop file somewhere to access its contents. Then a regular `cp` is enough to copy the contained files:
```console
sudo -s
mkdir /tmp/loop-src
mount -o ro /tmp/fstransform.mount.23404/.fstranform.loop.26458 /tmp/loop-src
cp -a /tmp/loop-src/. /tmp/weiwei2
```

**Comment by WeiweiDuan** [comment]
Thanks a lot for your prompt reply. I followed your instruction. But the loop-file becomes two loop files after copying to the disk. Please check the details in the following screenshot. Should I do the instruction again for the two loop files?
![image](https://user-images.githubusercontent.com/9730105/91749686-b8f5b480-eb76-11ea-965d-0e61f62db3a6.png)

**Comment by cosmos72** [comment]
It depends, there are a lot of files and mountpoints involved.
Which files are mounted where? The output of `mount` would help here.

**Comment by WeiweiDuan** [comment]
I mounted loop-file to '/data/loop-src' directory using "mount -o ro /tmp/fstransform.mount.23404/.fstransform.loop.26458 /data/loop-src/". Then copy files in /data/loop-src to '/tmp/weiwei' using "cp -av /data/loop-src/. /tmp/weiwei". Here is the output of mount:
![image](https://user-images.githubusercontent.com/9730105/91751915-60c0b180-eb7a-11ea-9355-e2d34f6fdeba.png)
![image](https://user-images.githubusercontent.com/9730105/91751935-69b18300-eb7a-11ea-84ad-b0fe4c68244e.png)

**Comment by cosmos72** [comment]
Then yes: do the instruction again for the two loop files.

This is because the conversion was attempted again (at least two times) on a half converted disk

**Comment by WeiweiDuan** [comment]
I have one more question. When I do "cp -av /tmp/loop-src/. /tmp/weiwei2", should the disk mounted in "tmp/weiwei2" be empty. Or can I create a directory and copy the content in /tmp/loop-src into it? Thanks for your help!

**Comment by cosmos72** [comment]
No, there is no need for the destination disk of `cp -av SOURCE DESTINATION` to be empty.
You can create a directory and copy the content from `/tmp/loop-src` into it.

The only caveat is: be careful not to **fill** the destination disk (`watch df` in another terminal is your friend)

**Comment by WeiweiDuan** [comment]
I mount the .fstransform.loop.22262 to '/data/loop-src2' and copied '/data/loop-src2' to '/data/loop-recovery'. But there still the loop-file in the loop-recovery. Should it be the original files?
![image](https://user-images.githubusercontent.com/9730105/91758444-e5183200-eb84-11ea-82d2-d5a3b7a2c10b.png)

**Comment by cosmos72** [comment]
Some of the `.fstransform.loop.XXYYZ` files may be empty - it happens if the conversion was interrupted very early.
You can recognize them because the command `file` will say they are `data`. For example:
```console
$ file .fstransform.loop.22262
.fstransform.loop.22262: data
```

If the file contains a mountable file system, it will be detected:
```console
$ file .fstransform.loop.22262
.fstransform.loop.22263: Linux rev 1.0 ext4 filesystem data, UUID=36f7e75a-0b26-4076-8dc6-cc31ffaf7184 (extents) (64bit) (large files) (huge files)
```

**Comment by WeiweiDuan** [comment]
Thanks for your reply. It turns out that the loop-file is mountable, but when I mounted it, there is an new error. Could you help me check about it? Really thanks for your help.
Here is the screenshot shown that the loop-file is mountable
![image](https://user-images.githubusercontent.com/9730105/92354705-aa882b00-f097-11ea-9109-4caa10e0f42a.png)
Here is the new error:
![image](https://user-images.githubusercontent.com/9730105/92354741-b96edd80-f097-11ea-8b16-2267f04d8ef7.png)

**Comment by cosmos72** [comment]
You can run a filesystem check, to detect and (where possible) fix errors:
```Console
sudo e2fsck /data/loop-src/.fstransform.loop.2262
```
To fix errors, the file must be inside a writeable file-system (not read-only).

**Comment by WeiweiDuan** [comment]
Thanks for your response. I did mount and copy successfully. But when I checked the file type of the file after 'cp' command, it is data. What I want is the original files, which should be a directory. Can I convert the intermediate converted files to the original files (the directory)? Thanks for your help.

**Comment by cosmos72** [comment]
If `mount` fails and `file` says it's raw data, there is not much you can do: it's very likely that the file contains nothing useful.
